subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/filter/TestSelectionMatcher.java

```diff
@@ -56,35 +64,77 @@ public TestSelectionMatcher(Collection<String> includedTests, Collection<String>
 
     public boolean matchesTest(String className, String methodName) {
         return matchesPattern(buildScriptIncludePatterns, className, methodName)
-            && matchesPattern(commandLineIncludePatterns, className, methodName);
+            && matchesPattern(commandLineIncludePatterns, className, methodName)
+            && !matchesExcludePattern(className, methodName);
     }
 
     public boolean mayIncludeClass(String fullQualifiedClassName) {
         return mayIncludeClass(buildScriptIncludePatterns, fullQualifiedClassName)
-            && mayIncludeClass(commandLineIncludePatterns, fullQualifiedClassName);
+            && mayIncludeClass(commandLineIncludePatterns, fullQualifiedClassName)
+            && !mayExcludeClass(fullQualifiedClassName);
     }
 
     private boolean mayIncludeClass(List<TestPattern> includePatterns, String fullQualifiedName) {
         if (includePatterns.isEmpty()) {
             return true;
         }
-        for (TestPattern pattern : includePatterns) {
+        return isClassNameMatch(includePatterns, fullQualifiedName);
+    }
+
+    private boolean mayExcludeClass(String fullQualifiedName) {
+        if (buildScriptExcludePatterns.isEmpty()) {
+            return false;
+        }
+        return isExactClassNameMatch(buildScriptExcludePatterns, fullQualifiedName);
+    }
+
+    private boolean isExactClassNameMatch(List<TestPattern> patterns, String fullQualifiedName) {
+        for (TestPattern pattern : patterns) {
+            if (pattern.matchesClass(fullQualifiedName)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isClassNameMatch(List<TestPattern> patterns, String fullQualifiedName) {
+        for (TestPattern pattern : patterns) {
             if (pattern.mayIncludeClass(fullQualifiedName)) {
                 return true;
             }
         }
         return false;
     }
 
-    private boolean matchesPattern(List<TestPattern> includePatterns, String fullQualifiedName, String methodName) {
+    private boolean matchesPattern(List<TestPattern> includePatterns, String className,
+        String methodName) {
         if (includePatterns.isEmpty()) {
             return true;
         }
-        for (TestPattern pattern : includePatterns) {
-            if (pattern.matchesClassAndMethod(fullQualifiedName, methodName)) {
+        return matchesIncludeOrExcludePattern(includePatterns, className, methodName);
+    }
+
+    private boolean matchesExcludePattern(String className, String methodName) {
+        if (buildScriptExcludePatterns.isEmpty()) {
+            return false;
+        }
+        if (isClassNameMatch(buildScriptExcludePatterns, className) && methodName == null) {
```

Comment:
I don't understand this part and why it is different from the include code path. Can you please explain with an example?

Reply:
@felipecsl Could you give me some feedback on this one please?